### PR TITLE
feat: Use labels on PVCs for better reconciling

### PR DESCRIPTION
**Pull request by mkuznyetsov**
### What does this PR do?
Add new labels to indicate PVCs that are in use by Devworkspace Operator.
This allows to trigger workspace reconcile, that must be triggered upon PVC deletion.
and if it's a common PVC that had a name configured by external operator config, only those workspaces that would have the same PVC name should be reconciled 

### What issues does this PR fix or reference?
https://github.com/devfile/devworkspace-operator/issues/920

### Is it tested? How?
Steps to test the new & old behavior with sample DevWorkspaces (DW)

1) install & run a **current version** of DWO:

https://github.com/devfile/devworkspace-operator/?tab=readme-ov-file#run-controller-locally

2) add some Devworkspaces and a devworkspace:

- DW with per-user storage

```
cat <<'EOF' | kubectl apply -n $NAMESPACE -f - 
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: old-dw-per-user
spec:
  started: true
  routingClass: 'basic'
  template:
    components:
      - name: web-terminal
        container:
          image: quay.io/wto/web-terminal-tooling:next
          memoryRequest: 256Mi
          memoryLimit: 512Mi
          mountSources: true
          command:
           - "tail"
           - "-f"
           - "/dev/null"
EOF
```

- DW with per-workspace storage
```
cat <<'EOF' | kubectl apply -n $NAMESPACE -f - 
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: old-dw-per-workspace
spec:
  started: true
  routingClass: 'basic'
  template:
    attributes:
      controller.devfile.io/storage-type: per-workspace
    components:
      - name: basic-volume
        volume: 
          size: 5Gi
      - name: web-terminal
        container:
          image: quay.io/wto/web-terminal-tooling:next
          memoryRequest: 256Mi
          memoryLimit: 512Mi
          mountSources: true
          command:
           - "tail"
           - "-f"
           - "/dev/null"
EOF
```

- External operator config, that is configured with custom PVC name

```
cat <<'EOF' | kubectl apply -f - 
apiVersion: controller.devfile.io/v1alpha1
config:
  enableExperimentalFeatures: true
  routing:
    clusterHostSuffix: 192.168.49.2.nip.io
    defaultRoutingClass: basic
  workspace:
    defaultStorageSize:
      common: 6Gi
      perWorkspace: 4Gi
    imagePullPolicy: Always
    pvcName: old-custom-pvc-name
kind: DevWorkspaceOperatorConfig
metadata:
  name: external-config
  namespace: default
EOF

cat <<'EOF' | kubectl apply -n $NAMESPACE -f - 
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: old-dw-per-user-external
spec:
  started: true
  routingClass: 'basic'
  template:
    attributes:
      controller.devfile.io/devworkspace-config:
        name: external-config
        namespace: default
      controller.devfile.io/storage-type: common
    components:
      - name: web-terminal
        container:
          image: quay.io/wto/web-terminal-tooling:next
          memoryRequest: 256Mi
          memoryLimit: 512Mi
          mountSources: true
          command:
           - "tail"
           - "-f"
           - "/dev/null"
EOF
```
  
Inspect PVC objects, that are created for these repos. 

```
kubectl describe pvc -n $NAMESPACE | grep "Labels:" -A 3 -B 6
```
- They WILL NOT have the label "controller.devfile.io/devworkspace_pvc_type". 
- Only per-workspace PVC will have "controller.devfile.io/devworkspace_id" label

1) build and run a **new version** of operator from this pull-request:
export NAMESPACE="devworkspace-controler"
...
make run

1) Create DWs from step TODO:

- DW with per-user storage

```
cat <<'EOF' | kubectl apply -n $NAMESPACE -f - 
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: new-dw-per-user
spec:
  started: true
  routingClass: 'basic'
  template:
    components:
      - name: web-terminal
        container:
          image: quay.io/wto/web-terminal-tooling:next
          memoryRequest: 256Mi
          memoryLimit: 512Mi
          mountSources: true
          command:
           - "tail"
           - "-f"
           - "/dev/null"
EOF
```

- DW with per-workspace storage
```
cat <<'EOF' | kubectl apply -n $NAMESPACE -f - 
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: new-dw-per-workspace
spec:
  started: true
  routingClass: 'basic'
  template:
    attributes:
      controller.devfile.io/storage-type: per-workspace
    components:
      - name: basic-volume
        volume: 
          size: 5Gi
      - name: web-terminal
        container:
          image: quay.io/wto/web-terminal-tooling:next
          memoryRequest: 256Mi
          memoryLimit: 512Mi
          mountSources: true
          command:
           - "tail"
           - "-f"
           - "/dev/null"
EOF
```

- Another external operator config, that is configured with custom PVC name

```
cat <<'EOF' | kubectl apply -f - 
apiVersion: controller.devfile.io/v1alpha1
config:
  enableExperimentalFeatures: true
  routing:
    clusterHostSuffix: 192.168.49.2.nip.io
    defaultRoutingClass: basic
  workspace:
    defaultStorageSize:
      common: 20Gi
      perWorkspace: 29Gi
    imagePullPolicy: Always
    pvcName: new-custom-pvc-name
kind: DevWorkspaceOperatorConfig
metadata:
  name: new-external-config
  namespace: default
EOF


cat <<'EOF' | kubectl apply -n $NAMESPACE -f - 
kind: DevWorkspace
apiVersion: workspace.devfile.io/v1alpha2
metadata:
  name: new-dw-per-user-external
spec:
  started: true
  routingClass: 'basic'
  template:
    attributes:
      controller.devfile.io/devworkspace-config:
        name: new-external-config
        namespace: default
      controller.devfile.io/storage-type: common
    components:
      - name: web-terminal
        container:
          image: quay.io/wto/web-terminal-tooling:next
          memoryRequest: 256Mi
          memoryLimit: 512Mi
          mountSources: true
          command:
           - "tail"
           - "-f"
           - "/dev/null"
EOF
```

```
kubectl describe pvc -n $NAMESPACE | grep "Labels:" -A 3 -B 6
```

Observe a label "controller.devfile.io/devworkspace_pvc_type" added to certain kubernetes PVC objects:

- `claim-devworkspace` PVC will not have this label, since this PVC was created with old DWO. Even new DWO
- `old-custom-pvc-name` PVC will not have this label, but `new-custom-pvc-name` will have it
- same for per-workspace, only the PVC for workspace started with new operator will have the type label

Now start deleting PVCs one after another with `kubectl delete pvc`
And see that workspaces are reconciled respectivelly.

### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

### CODE REVIEW

File: controllers/workspace/eventhandlers.go [hunk 1]
```diff
@@ -47,44 +47,63 @@ func dwRelatedPodsHandler(obj client.Object) []reconcile.Request {
...
-		if ownerref.Kind != "DevWorkspace" {
-			continue
-		}
-		return []reconcile.Request{
-			{
-				NamespacedName: types.NamespacedName{
-					Name:      ownerref.Name,
-					Namespace: obj.GetNamespace(),
+	// No need to reconcile PVC, if it doesn't have a label specifying its type
+	pvcLabel, ok := obj.GetLabels()[constants.DevWorkspacePVCTypeLabel]
```
Comment (dkwon17) [inline controllers/workspace/eventhandlers.go:51]: If a user upgrades from an older version of DWO to the latest version, will the older common PVCs get the new label?

File: controllers/workspace/eventhandlers.go [hunk 2]
```diff
@@ -47,44 +47,63 @@ func dwRelatedPodsHandler(obj client.Object) []reconcile.Request {
...
-	for _, ownerref := range obj.GetOwnerReferences() {
-		if ownerref.Kind != "DevWorkspace" {
-			continue
-		}
-		return []reconcile.Request{
-			{
-				NamespacedName: types.NamespacedName{
-					Name:      ownerref.Name,
-					Namespace: obj.GetNamespace(),
+	// No need to reconcile PVC, if it doesn't have a label specifying its type
```
Comment (dkwon17) [inline controllers/workspace/eventhandlers.go:50]:
```suggestion
	// No need to reconcile, if PVC doesn't have a label specifying its type
```

File: controllers/workspace/eventhandlers.go [hunk 3]
```diff
@@ -47,44 +47,63 @@ func dwRelatedPodsHandler(obj client.Object) []reconcile.Request {
...
+		if err := r.Client.List(context.Background(), dwList, &client.ListOptions{Namespace: obj.GetNamespace()}); err != nil {
+			return []reconcile.Request{}
 		}
+		var reconciles []reconcile.Request
+
+
+		for _, workspace := range dwList.Items {
+			// determin if PVC belongs to the workspace that has to be reconcile by its name
+			// which comes from the DW operator global, or external config that may be defined
+			// for this workspace
```
Comment (dkwon17) [inline controllers/workspace/eventhandlers.go:83]: I suggest:
```
Determine workspaces to reconcile that use the current common PVC. Workspaces can either use the common PVC where the PVC name is coming from the global config, or from an external config the workspace might use
```

File: controllers/workspace/eventhandlers.go [hunk 4]
```diff
@@ -47,44 +47,63 @@ func dwRelatedPodsHandler(obj client.Object) []reconcile.Request {
...
+			var workspacePvcName string;
+
+			if (workspace.Spec.Template.Attributes.Exists(constants.ExternalDevWorkspaceConfiguration)) {
```
Comment (dkwon17) [inline controllers/workspace/eventhandlers.go:86]: Should we check if the workspace uses common/per-user storage by checking 

`storageType := workspace.Spec.Template.Attributes.GetString(constants.DevWorkspaceStorageTypeAttribute, nil)`

first?

File: controllers/workspace/eventhandlers.go (hunk 1 above)
Comment (AObuchow) [reply controllers/workspace/eventhandlers.go:51]: @mkuznyetsov this is also my concern. There was an older [PR](https://github.com/devfile/devworkspace-operator/pull/1174) that added labels per-workspace PVCs, but I'm almost certain this only impacted _new_ per-workspace PVCs.

My concern stems from 2 things:

## 1. Making sure existing PVCs won't be deleted, then re-created to apply the label.
This would cause user data loss. Looking at the sync code, it seems like PVCs are [not considered mutable](https://github.com/devfile/devworkspace-operator/blob/main/pkg/provision/sync/sync.go#L183), so the new label would not get applied to existing PVCs. There is however a note that we could still allow for [updating labels](https://github.com/devfile/devworkspace-operator/blob/main/pkg/provision/sync/sync.go#L59).

If we want to update existing PVCs with the label, this complicates this PR further: either the sync code needs to be updated to only allow modifying PVCs if the labels are modified, or we need to manually `diff` the PVC after calling `SyncObjectWithCluster()`, see that the labels are missing and call` sync.updateObjectGeneric()` to apply the labels. The former approach is much more clean, though a bit trickier as `SyncObjectWithCluster()` is a critical function called throughout DWO. 

## 2. If we decide to not update existing PVCs with the new label, then the `dwPVCHandler()` should still work on PVCs that are missing the new label. 
This is currently not the case with the following check:

```go
	// No need to reconcile, if PVC doesn't have a label specifying its type
	pvcLabel, ok := obj.GetLabels()[constants.DevWorkspacePVCTypeLabel]

	if !ok {
		return []reconcile.Request{}
	}
```

You'd have to change the `dwPVCHandler` code to not assume that the `constants.DevWorkspacePVCTypeLabel` is applied to the PVC in question. If it does not exist, the behaviour should match the behaviour prior to this PR (i.e. the global common PVC name is assumed to be used, and the original PR bug will be present). If the label does exist, then you can make use of the logic that you've implemented. 

---

In general, I think it's easier as a first step to have this PR _not_ update existing PVCs with the new label (i.e. consider my 2nd concern). A later PR could update the sync code to allow label updating on PVCs.

File: pkg/constants/metadata.go [hunk 5]
```diff
@@ -20,6 +20,9 @@ const (
 	// DevWorkspaceIDLabel is the label key to store workspace identifier
 	DevWorkspaceIDLabel = "controller.devfile.io/devworkspace_id"
 
+	// DevWorkspacePVCTypeLabel is the label key to identify DeWorkspace common PVCs and their type
```
Comment (AObuchow) [inline pkg/constants/metadata.go:23]: This comment is slightly inaccurate, as we are using this label for both per-workspace and common/per-user PVC's.

I would rewrite it as:
`// DevWorkspacePVCTypeLabel is the label key to identify PVCs used by DevWorkspaces and indicate their storage strategy.`

**Review by AObuchow — CHANGES_REQUESTED** [review]
@mkuznyetsov Thank you for the great first PR :partying_face: Excited to have you contributing to the project. I've made a few initial remarks, but overall things are looking good :)

File: controllers/workspace/eventhandlers.go (hunk 1 above)
Comment (mkuznyetsov) [reply controllers/workspace/eventhandlers.go]: alright, I will rewrite to account for the use of PVC with missing label, and leave a reference to this PR why this has to be this way

File: controllers/workspace/eventhandlers.go (hunk 4 above)
Comment (mkuznyetsov) [reply controllers/workspace/eventhandlers.go]: yes, thanks

**Comment by AObuchow** [comment]
@mkuznyetsov sorry about the failing CI tests, that's due to https://github.com/devfile/devworkspace-operator/issues/1242 (not related to your PR). I've opened https://github.com/devfile/devworkspace-operator/pull/1243 to resolve this.

File: controllers/workspace/eventhandlers.go [hunk 6]
```diff
@@ -47,44 +47,63 @@ func dwRelatedPodsHandler(obj client.Object) []reconcile.Request {
...
+				externalConfig, _ := wkspConfig.ResolveConfigForWorkspace(&workspace, r.Client)
+				workspacePvcName = externalConfig.Workspace.PVCName
+			}
+			if len(workspacePvcName) == 0 {
```
[the remark this answers is not in the document]
Comment (mkuznyetsov) [reply controllers/workspace/eventhandlers.go]: fixed

File: controllers/workspace/eventhandlers.go [hunk 7]
```diff
@@ -47,44 +47,63 @@ func dwRelatedPodsHandler(obj client.Object) []reconcile.Request {
...
+
+
+		for _, workspace := range dwList.Items {
+			// determin if PVC belongs to the workspace that has to be reconcile by its name
+			// which comes from the DW operator global, or external config that may be defined
+			// for this workspace
+			var workspacePvcName string;
+
+			if (workspace.Spec.Template.Attributes.Exists(constants.ExternalDevWorkspaceConfiguration)) {
+				externalConfig, _ := wkspConfig.ResolveConfigForWorkspace(&workspace, r.Client)
```
[the remark this answers is not in the document]
Comment (mkuznyetsov) [reply controllers/workspace/eventhandlers.go]: added logger

File: pkg/constants/metadata.go (hunk 5 above)
Comment (mkuznyetsov) [reply pkg/constants/metadata.go]: thx, comment applied

File: controllers/workspace/eventhandlers.go (hunk 1 above)
Comment (mkuznyetsov) [reply controllers/workspace/eventhandlers.go]: pr rewritten to support old usecase, when there is no new label

File: controllers/workspace/eventhandlers.go (hunk 3 above)
Comment (mkuznyetsov) [reply controllers/workspace/eventhandlers.go]: comment applied

File: controllers/workspace/eventhandlers.go [hunk 8]
```diff
@@ -47,7 +47,13 @@ func dwRelatedPodsHandler(obj client.Object) []reconcile.Request {
 }
 
 func (r *DevWorkspaceReconciler) dwPVCHandler(obj client.Object) []reconcile.Request {
-	// Check if PVC is owned by a DevWorkspace (per-workspace storage case)
+	if obj.GetDeletionTimestamp() == nil {
+		return []reconcile.Request{}
+	}
+	// we can wrap the code below for handling per-workspace PVCs
```
Comment (AObuchow) [inline controllers/workspace/eventhandlers.go:53]: Thanks for reporting https://github.com/devfile/devworkspace-operator/issues/1250 -- I think this comment could be now simplified to something along the lines of:

```
// TODO: Ensure all new and existing PVC's get the `controller.devfile.io/devworkspace_pvc_type` label.
// See: https://github.com/devfile/devworkspace-operator/issues/1250
```

File: pkg/constants/metadata.go [hunk 9]
```diff
@@ -20,6 +20,9 @@ const (
 	// DevWorkspaceIDLabel is the label key to store workspace identifier
 	DevWorkspaceIDLabel = "controller.devfile.io/devworkspace_id"
 
+	// DevWorkspacePVCTypeLabel is the label key to identify PVCs used by DevWorkspaces and indicate their storage strategy.
+	DevWorkspacePVCTypeLabel = "controller.devfile.io/devworkspace_pvc_type_label"
```
Comment (AObuchow) [inline pkg/constants/metadata.go:24]: I think the actual string value of this label could just be `controller.devfile.io/devworkspace_pvc_type` (i.e. remove the `_label` part at the end), since it's implied by its usage as a label in the PVC spec, and its name `DevWorkspacePVCTypeLabel` in the code.

File: controllers/workspace/eventhandlers.go [hunk 10]
```diff
@@ -47,7 +47,13 @@ func dwRelatedPodsHandler(obj client.Object) []reconcile.Request {
 }
 
 func (r *DevWorkspaceReconciler) dwPVCHandler(obj client.Object) []reconcile.Request {
-	// Check if PVC is owned by a DevWorkspace (per-workspace storage case)
+	if obj.GetDeletionTimestamp() == nil {
+		return []reconcile.Request{}
```
Comment (AObuchow) [inline controllers/workspace/eventhandlers.go:51]: I'm slightly hesitant to prevent reconciles unless the PVC was deleted (i.e. ignoring modifications to a per-workspace PVC), just because @amisevsk might have had some edge case in mind when he made https://github.com/devfile/devworkspace-operator/commit/2a44ff6da5aa079edd1f21b4032bb9aba10600c2:

> Update the watch for PVCs to also enqueue reconciles for per-workspace
> PVC events. Previously, the per-workspace PVC could be **modified**/deleted
> without the DevWorkspace that owns that PVC noticing.

However:
- We weren't queuing reconciles for common/per-user PVCs unless they were being deleted, so there's an inconsistency between per-workspace PVCs & per-user PVCs causing reconciles upon modification
- Users are **not** supposed to **modify** PVCs created by DWO. Attempting to do so in the OpenShift UI will present the user with a warning: `This resource is managed by DevWorkspaceDW
<DevWorkspace Name> and any modifications may be overwritten. Edit the managing resource to preserve changes.`. **However:** editing the PVC with kubectl/oc will **not give any warning.**
- As discussed in point 1. of my earlier comment https://github.com/devfile/devworkspace-operator/pull/1233#discussion_r1508110341: PVCs are currently considered immutable from DWO's perspective. Any modification made to a DWO-owned PVC cannot be automatically amended by DWO, so reconciling on modification doesn't currently yield any advantages. **However:** in the future, if we resolve https://github.com/devfile/devworkspace-operator/issues/1250 and allow for DWO to modify PVCs to add the `controller.devfile.io/devworkspace_pvc_type` label, then a user removing this label _should_ trigger a reconcile to allow DWO to re-apply the label. This is a problem for the future however...


In summary, I think this optimization is okay :man_shrugging: I just wanted to write out these thoughts beforehand.

File: controllers/workspace/eventhandlers.go [hunk 11]
```diff
@@ -47,7 +47,13 @@ func dwRelatedPodsHandler(obj client.Object) []reconcile.Request {
 }
 
 func (r *DevWorkspaceReconciler) dwPVCHandler(obj client.Object) []reconcile.Request {
-	// Check if PVC is owned by a DevWorkspace (per-workspace storage case)
```
Comment (AObuchow) [inline controllers/workspace/eventhandlers.go:50]: Rather than removing this comment, I'd move it back on-top of the code it was [originally above](https://github.com/devfile/devworkspace-operator/pull/1233/files#diff-1c6fd18e74606af2c4edc2a26c6df20b2030af8b9f72e3c6bb28796ee11620bdR57)

File: controllers/workspace/eventhandlers.go [hunk 12]
```diff
@@ -62,20 +68,41 @@ func (r *DevWorkspaceReconciler) dwPVCHandler(obj client.Object) []reconcile.Req
...
-		if storageType == constants.CommonStorageClassType || storageType == constants.PerUserStorageClassType || storageType == "" {
+		//Determine workspaces to reconcile that use the current common PVC.
+		// Workspaces can either use the common PVC where the PVC name
+		// is coming from the global config, or from an external config the workspace might use
+		workspacePVCName := wkspConfig.GetGlobalConfig().Workspace.PVCName
+
+		if workspace.Spec.Template.Attributes.Exists(constants.ExternalDevWorkspaceConfiguration) {
+			externalConfig, err := wkspConfig.ResolveConfigForWorkspace(&workspace, r.Client)
+			if err != nil {
+				r.Log.Info("Couldn't fetch external config for workspace %s, using PVC Name from global config instead", err.Error())
```
Comment (AObuchow) [inline controllers/workspace/eventhandlers.go:97]: We should test this logger works as expected by:
1. Creating an external DWOC that uses a custom PVC name
2. Using the external DWOC in a workspace that uses the common PVC storage strategy
3. Starting up the workspace that uses the external DWOC, wait for it to be in the READY state
4. Deleting the external DWOC
5. Deleting the common PVC
6. Ensure the log appears in DWO's logs

**Comment by AObuchow** [comment]
@mkuznyetsov when you get a chance, please rebase your PR onto the main branch so that the Validate PR GH Action will pass btw :)

File: controllers/workspace/eventhandlers.go (hunk 12 above)
Comment (mkuznyetsov) [reply controllers/workspace/eventhandlers.go:97]: Example logger output, if devworkpace had it external config deleted, when attempting to reconcile
`
{"level":"info","ts":"2024-04-10T18:39:02+03:00","logger":"controllers.DevWorkspace","msg":"Couldn't resolve PVC name for workspace 'external-plain1' in namespace 'devworkspace-controller', using PVC name 'claim-devworkspace' from global config instead: could not fetch external DWOC with name ec1 in namespace default: DevWorkspaceOperatorConfig.controller.devfile.io \"ec1\" not found."}
`

**Review by AObuchow — APPROVED** [review]
@mkuznyetsov overall looks good to me :) Manual testing was complicated by https://github.com/devfile/devworkspace-operator/issues/1258, but your PR seems good to me. 

Let's wait until I get back from PTO on May 9th before merging this.

Amazing work & congrats on your first (soon to be finally merged!!) PR :D

**Comment by AObuchow** [comment]
@mkuznyetsov I think this PR is good to merge :) Before we do so, can you squash your commits with a `git rebase --autosquash` please? I also think [`a705e11` (#1233)](https://github.com/devfile/devworkspace-operator/pull/1233/commits/a705e118fc784cd0e3274689412bf7d0050055e8) could be squashed into [`08e12bc` (#1233)](https://github.com/devfile/devworkspace-operator/pull/1233/commits/08e12bc0b5938b82766d2bef191e64af9f5a7873).

Additionally, the commit description of [`08e12bc` (#1233)](https://github.com/devfile/devworkspace-operator/pull/1233/commits/08e12bc0b5938b82766d2bef191e64af9f5a7873) could maybe be re-worded to something like: `feat: label PVCs by storage strategy`

File: controllers/workspace/eventhandlers.go [hunk 13]
```diff
@@ -62,20 +70,42 @@ func (r *DevWorkspaceReconciler) dwPVCHandler(obj client.Object) []reconcile.Req
...
+		// is coming from the global config, or from an external config the workspace might use
+		workspacePVCName := wkspConfig.GetGlobalConfig().Workspace.PVCName
+
+		if workspace.Spec.Template.Attributes.Exists(constants.ExternalDevWorkspaceConfiguration) {
+			externalConfig, err := wkspConfig.ResolveConfigForWorkspace(&workspace, r.Client)
+			if err != nil {
+				r.Log.Info(fmt.Sprintf("Couldn't resolve PVC name for workspace '%s' in namespace '%s', using PVC name '%s' from global config instead: %s.", workspace.Name, workspace.Namespace, workspacePVCName, err.Error()))
+			} else {
+				storageType := workspace.Spec.Template.Attributes.GetString(constants.DevWorkspaceStorageTypeAttribute, nil)
+				if storageType == constants.CommonStorageClassType || storageType == constants.PerUserStorageClassType {
```
Comment (dkwon17) [inline controllers/workspace/eventhandlers.go:103]: Do we need to also check `storageType == ""` here?

Comment (AObuchow) [reply controllers/workspace/eventhandlers.go:103]: @dkwon17 good catch!! Yes, we should be checking this too @mkuznyetsov

Comment (dkwon17) [reply controllers/workspace/eventhandlers.go:103]: Also, should we be doing this check in the first line of this for loop instead?

Comment (dkwon17) [reply controllers/workspace/eventhandlers.go:103]: Because AFAIK we are looking for common/per-user workspaces to reconcile in this for loop

Comment (AObuchow) [reply controllers/workspace/eventhandlers.go:103]: +1 @dkwon17 you're right, the order of checking for workspace storage strategy before iterating over a workspace should have not changed. @mkuznyetsov please restore the [2 lines](https://github.com/devfile/devworkspace-operator/pull/1233/files#diff-1c6fd18e74606af2c4edc2a26c6df20b2030af8b9f72e3c6bb28796ee11620bdL77-L78) that were moved for checking the workspace storage strategy, so that they occur before we retrieve the PVC name.

File: pkg/provision/storage/shared.go [hunk 14]
```diff
@@ -61,7 +61,6 @@ func WorkspaceNeedsStorage(workspace *dw.DevWorkspaceTemplateSpec) bool {
 }
 
 func getPVCSpec(name, namespace string, storageClass *string, size resource.Quantity) (*corev1.PersistentVolumeClaim, error) {
-
```
Comment (AObuchow) [inline pkg/provision/storage/shared.go:64]: @mkuznyetsov sorry for nit-pick, but could you please also restore this newline so that the diff is a bit cleaner?

Comment (AObuchow) [reply pkg/provision/storage/shared.go:64]: If possible, please re-add this newline.

File: pkg/provision/storage/shared.go [hunk 15]
```diff
@@ -61,7 +61,7 @@ func WorkspaceNeedsStorage(workspace *dw.DevWorkspaceTemplateSpec) bool {
...
+
```
Comment (AObuchow) [inline pkg/provision/storage/shared.go:64]: If possible, please remove this tab (unless it's coming from go format)

**Review by AObuchow — COMMENTED** [review]
Please run `make format` with goimports installed `go install golang.org/x/tools/cmd/goimports@latest`

File: pkg/provision/storage/shared.go (hunk 15 above)
Comment (mkuznyetsov) [reply pkg/provision/storage/shared.go]: oops, fixed it in the amended commit

**Review by AObuchow — APPROVED** [review]
LGTM :)

**Comment by AObuchow** [comment]
Congrats on having your first PR merged to the project @mkuznyetsov :tada: Amazing work :)
